lib/stop_context.go

```diff
@@ -0,0 +1,37 @@
+package lib
+
+import (
+	"context"
+	"time"
+)
+
+// StopChannelContext implements the context.Context interface
+// You provide the channel to select on to determine whether
+// the context should be cancelled and other code such
+// as the rate.Limiter will automatically use the channel
+// appropriately
+type StopChannelContext struct {
```

Comment:
What's the value in this over just using `context.Context`? I nearly called that out in a few PRs recently that are using `stopCh` pattern in new places. I'm in two minds myself but it seems like it might be cleaner to just write new code to use contexts "properly" rather than add a shim so we can still use plain stop chans and then pass them into context-aware code?

This is simple enough if it serves a purpose for legacy code we don't want to refactor though, just not sure if it's encouraging writing context-unaware code and then using this shim which seems like a bad move for the long term?

Reply:
Well none really other than it allows us to use a "Context" when we already have a stopCh. I will investigate moving to using context all over instead. It shouldn't be too hard.